sqlglot/expressions.py

```diff
@@ -4462,6 +4462,10 @@ class ToChar(Func):
     arg_types = {"this": True, "format": False, "nlsparam": False}
 
 
+class ToNumber(Func):
+    arg_types = {"this": True, "format": False, "nlsparam": False}
+
```

Comment:
When defining `Func`s, you need to take into account what other dialects do. This definition implicitly makes it so that `TO_NUMBER` is a function that expects at most three arguments for _every_ dialect. This is not necessarily the case though, for example Snowflake also [supports](https://docs.snowflake.com/en/sql-reference/functions/to_decimal) `to_number(expr, fmt, precision, scale)`, so this PR would break queries that use this function with four arguments. So you need to take this into account and make sure `ToNumber` can represent Snowflake's variant as well, or fall back to parsing it as `Anonymous` if there are >2 arguments in Snowflake.

Additionally right now the base generator converts `ToNumber` into a cast. This will also break valid queries that use `TO_NUMBER` with a format in snowflake, postgres, and various other dialects that support this function. This needs to also be fixed if you're gonna introduce this expression.

With a quick look, these are the dialects that support `TO_NUMBER` and the respective supported arguments:

```
to_number(expr, fmt)
databricks, drill, postgres, redshift, spark

to_number(expr, fmt, nlsparam)
oracle, teradata 

to_number(expr, fmt, precision, scale)
snowflake
```

Please refactor this PR so that it handles these cases & make sure that it's properly tested. At the very minimum we need to ensure that a query containing `TO_NUMBER` in these dialects can be roundtripped (e.g. `parse_one` and `generate` with read=write should not change `TO_NUMBER`).

Reply:
I have made modifications, please review it